pykeops/torch/utils.py

```diff
@@ -157,6 +172,122 @@ def device(x):
         else:
             return None
 
+    @staticmethod
+    def distance_function(metric):
+        def euclidean(x, y):
+            return ((x - y) ** 2).sum(-1)
+
+        def manhattan(x, y):
+            return ((x - y).abs()).sum(-1)
+
+        def angular(x, y):
+            return -(x | y)
+
+        def angular_full(x, y):
+            return angular(x, y) / ((angular(x, x) * angular(y, y)).sqrt())
+
+        def hyperbolic(x, y):
+            return ((x - y) ** 2).sum(-1) / (x[0] * y[0])
+
+        if metric == "euclidean":
+            return euclidean
+        elif metric == "manhattan":
+            return manhattan
+        elif metric == "angular":
+            return angular
+        elif metric == "angular_full":
+            return angular_full
+        elif metric == "hyperbolic":
+            return hyperbolic
+        else:
+            raise ValueError("Unknown metric")
+
+    @staticmethod
+    def sort(x):
+        return torch.sort(x)
+
+    @staticmethod
+    def unsqueeze(x, n):
+        return torch.unsqueeze(x, n)
+
+    @staticmethod
+    def arange(n, device="cpu"):
+        return torch.arange(n, device=device)
+
+    @staticmethod
+    def repeat(x, n):
+        return torch.repeat_interleave(x, n)
+
+    @staticmethod
+    def to(x, device):
+        if isinstance(x, torch.Tensor):
+            return x.to(device)
+        return x
+
+    @staticmethod
+    def index_select(input, dim, index):
+        return torch.index_select(input, dim, index)
+
+    @staticmethod
+    def kmeans(x, distance=None, K=10, Niter=15, device="cuda", approx=False, n=10):
+
+        from pykeops.torch import LazyTensor
+
+        if distance is None:
+            distance = torchtools.distance_function("euclidean")
+
+        def calc_centroid(x, c, cl, n=10):
```

Comment:
I believe that we could do with a few comments here :-)

Reply:
done